File: icsm31-projeto/server-interpreted/linalg.py
"""
Algebra linear em Python puro — sem numpy/scipy.

Todos os "calculos" da reconstrucao (produto matriz-vetor, produto da
transposta por vetor, produto interno, norma e combinacoes lineares) sao
implementados aqui manualmente. Nenhuma biblioteca de calculo numerico e
utilizada: usa-se apenas o modulo `array` da biblioteca padrao como
contêiner denso de doubles (8 bytes por elemento), o que mantem o uso de
memoria viavel mesmo para a matriz H grande (ex.: 50816 x 3600).

A matriz e guardada em ordem row-major num unico `array('d')`:
    H[i][j] == data[i * cols + j]

Desempenho (ainda 100% Python puro, sem libs de calculo):
  - Os produtos internos usam `sum(map(mul, ...))`, que empurra o laco para
    o C do interpretador (map/sum), em vez de um `for` byte a byte.
  - `rmatvec` (H^T @ y) e feito como um matvec sobre a transposta H^T, que e
    construida uma unica vez e cacheada (`_HT`). Assim tanto H @ x quanto
    H^T @ y percorrem linhas contiguas — o caso rapido do padrao acima.
    Custo: uma copia extra da matriz em memoria (aceitavel, e paga so quando
    rmatvec e usado, e os servidores rodam um de cada vez).
"""

from __future__ import annotations

from array import array
from operator import mul
from typing import List, Optional, Sequence, Tuple


class Matrix:
    """Matriz densa de doubles armazenada num array plano (row-major)."""

    __slots__ = ("data", "rows", "cols", "_HT")

    def __init__(self, data: "array", rows: int, cols: int) -> None:
        self.data = data
        self.rows = rows
        self.cols = cols
        # Transposta (row-major) construida sob demanda em _ensure_transpose().
        self._HT: Optional["array"] = None

    @property
    def shape(self) -> Tuple[int, int]:
        return (self.rows, self.cols)

    def matvec(self, x: Sequence[float]) -> List[float]:
        """Retorna H @ x (vetor de tamanho `rows`)."""
        data = self.data
        cols = self.cols
        out = [0.0] * self.rows
        base = 0
        for i in range(self.rows):
            out[i] = sum(map(mul, data[base:base + cols], x))
            base += cols
        return out

    def _ensure_transpose(self) -> "array":
        """Constroi (uma vez) e cacheia H^T em ordem row-major.

        HT[j * rows + i] == H[i][j] == data[i * cols + j]
        """
        if self._HT is not None:
            return self._HT
        data = self.data
        rows = self.rows
        cols = self.cols
        HT = array("d", bytes(8 * rows * cols))
        base = 0
        for i in range(rows):
            # espalha a linha i de H na coluna i de H^T (passo `rows`)
            k = i
            for v in data[base:base + cols]:
                HT[k] = v
                k += rows
            base += cols
        self._HT = HT
        return HT

    def rmatvec(self, y: Sequence[float]) -> List[float]:
        """Retorna H^T @ y (vetor de tamanho `cols`).

        Implementado como um matvec sobre a transposta cacheada H^T.
        """
        HT = self._ensure_transpose()
        rows = self.rows
        out = [0.0] * self.cols
        base = 0
        for j in range(self.cols):
            out[j] = sum(map(mul, HT[base:base + rows], y))
            base += rows
        return out


def dot(a: Sequence[float], b: Sequence[float]) -> float:
    """Produto interno a . b (via map/sum, ainda Python puro)."""
    return sum(map(mul, a, b))


def norm(a: Sequence[float]) -> float:
    """Norma-2 (euclidiana) de a — sqrt via expoente 0.5, sem bibliotecas."""
    return sum(map(mul, a, a)) ** 0.5


def axpy(alpha: float, x: Sequence[float], y: Sequence[float]) -> List[float]:
    """Retorna alpha * x + y (novo vetor)."""
    return [alpha * xi + yi for xi, yi in zip(x, y)]


def sub(a: Sequence[float], b: Sequence[float]) -> List[float]:
    """Retorna a - b (novo vetor)."""
    return [ai - bi for ai, bi in zip(a, b)]


def load_matrix_csv(path: str) -> Matrix:
    """Carrega uma matriz densa de um arquivo CSV/texto, em Python puro.

    Cada linha do arquivo vira uma linha da matriz; os valores podem estar
    separados por virgula (padrao dos dados do professor) ou por espaco.
    """
    data = array("d")
    rows = 0
    cols = 0
    with open(path, "r") as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            parts = line.split(",") if "," in line else line.split()
            values = [float(p) for p in parts]
            if cols == 0:
                cols = len(values)
            elif len(values) != cols:
                raise ValueError(
                    f"linha {rows + 1} de {path} tem {len(values)} colunas, "
                    f"esperado {cols}"
                )
            data.extend(values)
            rows += 1
    return Matrix(data, rows, cols)


def load_vector_csv(path: str) -> List[float]:
    """Carrega um vetor (um valor por linha, ou linha unica separada) em Python puro."""
    values: List[float] = []
    with open(path, "r") as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            parts = line.split(",") if "," in line else line.split()
            values.extend(float(p) for p in parts)
    return values

File: icsm31-projeto/server-compiled/linalg.go
// Package main — Algebra linear em Go puro, sem gonum.
//
// Todos os "calculos" da reconstrucao (produto matriz-vetor, produto da
// transposta por vetor, produto interno, norma, raiz quadrada e valor absoluto)
// sao implementados aqui manualmente. NENHUMA biblioteca e utilizada — nem
// mesmo o pacote `math`. A matriz e guardada de forma densa num unico slice em
// ordem row-major:
//
//	H[i][j] == Data[i*Cols + j]
package main

// Matrix e uma matriz densa de float64 (row-major).
type Matrix struct {
	Data []float64
	Rows int
	Cols int
}

// NewMatrix cria uma matriz a partir de um slice plano row-major.
func NewMatrix(rows, cols int, data []float64) *Matrix {
	return &Matrix{Data: data, Rows: rows, Cols: cols}
}

// Dims devolve (linhas, colunas).
func (m *Matrix) Dims() (int, int) { return m.Rows, m.Cols }

// MatVec calcula H * x (vetor de tamanho Rows).
func (m *Matrix) MatVec(x []float64) []float64 {
	out := make([]float64, m.Rows)
	data := m.Data
	cols := m.Cols
	base := 0
	for i := 0; i < m.Rows; i++ {
		var s float64
		for j := 0; j < cols; j++ {
			s += data[base+j] * x[j]
		}
		out[i] = s
		base += cols
	}
	return out
}

// TMatVec calcula H^T * y (vetor de tamanho Cols).
func (m *Matrix) TMatVec(y []float64) []float64 {
	out := make([]float64, m.Cols)
	data := m.Data
	cols := m.Cols
	base := 0
	for i := 0; i < m.Rows; i++ {
		yi := y[i]
		if yi != 0 {
			for j := 0; j < cols; j++ {
				out[j] += data[base+j] * yi
			}
		}
		base += cols
	}
	return out
}

// Dot calcula o produto interno a . b.
func Dot(a, b []float64) float64 {
	var s float64
	for i := range a {
		s += a[i] * b[i]
	}
	return s
}

// Abs devolve o valor absoluto de x — implementado no proprio codigo (sem math).
func Abs(x float64) float64 {
	if x < 0 {
		return -x
	}
	return x
}

// Sqrt calcula a raiz quadrada por Newton-Raphson, no proprio codigo (sem math).
//
// Resolve x^2 = a iterando x <- (x + a/x) / 2 ate estabilizar. Para a <= 0
// retorna 0 (os usos deste projeto — normas e sqrt(l) do ganho — sao >= 0).
func Sqrt(a float64) float64 {
	if a <= 0 {
		return 0
	}
	x := a
	for i := 0; i < 100; i++ {
		nx := 0.5 * (x + a/x)
		if Abs(nx-x) <= 1e-15*nx {
			return nx
		}
		x = nx
	}
	return x
}

// Norm calcula a norma-2 (euclidiana) de a, usando o Sqrt implementado acima.
func Norm(a []float64) float64 {
	return Sqrt(Dot(a, a))
}
